### DISCUSSION TITLE
reportUnknownParameterType and 3rd party libraries

### DISCUSSION THREAD

**Discussion by hmc-cs-mdrissi** [microsoft/pyright#2376] [Q&A] [answered]
My current codebase works with a lot of untyped python libraries. The libraries are large enough (tensorflow, apache beam) that type hinting a meaningful portion of them looks fairly time consuming. At the same time I want all code within my team's codebase to be fully type hinted. The challenge is writing type hints like this,

```python
from untyped_library import Foo

def f(x: Foo) -> None
  ...
```

I want that to type check. I currently get a type error of unknown parameter type. Dropping that rule is undesirable when missing type for an argument/return I consider an error. I effectively want all types from untyped libraries to be treated as Any instead of unknown.

I did find one sorta workaround which was add a typings folder and add untyped_library/__init__.pyi with,

```python
from typing import Any


def __getattr__(name: str) -> Any:
    ...

```

This has one major downside. Before I added this while Foo was unknown I could double click on Foo and it'd jump to where it was defined. Now I add this and it loses that ide navigation. 

For mypy I'm able to have things work with configuration of,

```
disallow_untyped_defs = True

# Once per untyped library
ignore_missing_imports = True
```

The getattr solution is very close to that and from a CI perspective works. It worsens the IDE experience as it loses the original definition. Between losing IDE powers vs many type ignores neither looks good and I'm undecided which I'll pick.

For a concrete example I think any untyped library should work and explore import of navigation with typings. One specific one I've used,

`from apache_beam.pipeline import Pipeline`

and then trying typings/apache_beam/pipeline.pyi with getattr contents.

To install that library `pip install apache_beam==2.31.0`. I think my ideal solution here is either an option to make untyped library imports Any or reportUnknownParameterType to have a sibling option reportMissingParameterType. Right now,

```python
from untyped_library import Foo

def f(x) -> None
  ...
```

and

```python
from untyped_library import Foo

def f(x: Foo) -> None
  ...
```

are both type errors even though second is an improvement to me from the first. Some way to distinguish the two in reporting errors would be very helpful.

**Comment by erictraut** [comment, marked as answer]
I discussed this with my colleagues, and we think there's merit in adding a new diagnostic check for `reportMissingParameterType`. I'll  create an enhancement request that links to this discussion.

**Reply by erictraut** [reply]
Digging into this a bit deeper, have you tried enabling `reportUnknownParameterType` without any of the other `reportUnknown...` diagnostic rules? I think this might already provide you with the feature you're looking for.

**Reply by hmc-cs-mdrissi** [reply]
Testing it with minimal pyright config still gives me that error. For pylance it's needed to have  "useLibraryCodeForTypes": false to make result consistent with cli. "useLibraryCodeForTypes": true fixes it but useLibraryCodeForTypes creates a lot of false positives when working with a couple libraries (tensorflow is particularly an issue).

```
{
    "useLibraryCodeForTypes": false,
    "pythonVersion": "3.8",
    "reportUnknownParameterType": true
}
```

Is a minimal pyright config I'm using to see the error. The toy example I'm testing is

```
from apache_beam.pipeline import Pipeline


def f(x: Pipeline) -> None:
    ...
```

**Reply by erictraut** [reply]
Yep, you're right.

OK, then I think it makes sense to add a separate `reportMissingParameterType` check.

**Reply by erictraut** [reply]
I've added this new diagnostic check. It will be included in the next release of pyright.

**Reply by hmc-cs-mdrissi** [reply]
Thank you very much. This is a feature I know my team will enjoy and simplify a lot of current type ignores.
